miniapps/autodiff/admfem.hpp

```diff
@@ -0,0 +1,674 @@
+// Copyright (c) 2010-2021, Lawrence Livermore National Security, LLC. Produced
+// at the Lawrence Livermore National Laboratory. All Rights reserved. See files
+// LICENSE and NOTICE for details. LLNL-CODE-806117.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability visit https://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+
+#ifndef ADMFEM_HPP 
+#define ADMFEM_HPP
+
+#include "mfem.hpp"
+#include "fdual.hpp"
+#include "tadvector.hpp"
+#include "taddensemat.hpp"
+
+#ifdef MFEM_USE_CODIPACK
+#include <codi.hpp>
+
+namespace mfem {
+
+namespace ad {
+#ifdef MFEM_USE_ADFORWARD
+/// Forward AD type declaration
+typedef codi::RealForward ADFloatType;
+/// Vector type for AD-numbers
+typedef TADVector<ADFloatType> ADVectorType;
+/// Matrix type for AD-numbers
+typedef TADDenseMatrix<ADFloatType> ADMatrixType;
+#else
+/// Reverse AD type declaration
+typedef codi::RealReverse ADFloatType;
+/// Vector type for AD-numbers
+typedef TADVector<ADFloatType> ADVectorType;
+/// Matrix type for AD-numbers
+typedef TADDenseMatrix<ADFloatType> ADMatrixType;
+#endif
+}
+
+/// The class provides an evaluation of the Jacobian of a
+/// templated vector function provided in the constructor.
+/// The Jacobian is evaluated with the help of automatic
+/// differentiation (AD)
+/// https://en.wikipedia.org/wiki/Automatic_differentiation.
+/// The template parameters specify the size of the return
+/// vector (vector_size), the size of the input vector
+/// (state_size), and the size of the parameters supplied
+/// to the function.
+template<int vector_size=1, int state_size=1, int param_size=0>
+class VectorFuncAutoDiff
+{
+public:
+    /// F_ is user implemented function to be differentiated by VectorFuncAutoDiff.
+    /// The signature of the function is:
+    /// F_(mfem::Vector& parameters, ad::ADVectroType& state_vector, ad::ADVectorType& result).
+    /// The parameters vector should have size param_size. The state_vector should have size
+    /// state_size, and the result vector should have size vector_size. All size parameters are
+    /// teplate parameters in VectorFuncAutoDiff.
+    VectorFuncAutoDiff(std::function<void(mfem::Vector&, ad::ADVectorType&, ad::ADVectorType&)> F_)
+    {
+            F=F_;
+    }
+
+    void QJacobian(mfem::Vector &vparam, mfem::Vector &uu, mfem::DenseMatrix &jac)
+    {
+#ifdef MFEM_USE_ADFORWARD
+        // use forward mode
+        jac.SetSize(vector_size, state_size);
+        jac = 0.0;
+        {
+            ad::ADVectorType aduu(state_size);
```

Comment:
I'd just replace `uu` with `state` or similar; `aduu` is basically `ad_state` and `rr` is basically `ad_result`. More people will read this code than write it, so it's a net productivity loss to use such short variable names on non-index quantities because it forces people to spend more time deciphering code than doing science with it.

Reply:
I have replaced them on the particular location, however, I am afraid I might introduce additional bugs if I replace them everywhere in the code. For the upcoming updates with other AD libraries I will use ad_state and ad_result.